### PR TITLE
[new release] mirage-crypto (9 packages) (1.0.0)

### PR DESCRIPTION
**Pull request by hannesm**
Simple symmetric cryptography for the modern age

- Project page: <a href="https://github.com/mirage/mirage-crypto">https://github.com/mirage/mirage-crypto</a>
- Documentation: <a href="https://mirage.github.io/mirage-crypto/doc">https://mirage.github.io/mirage-crypto/doc</a>

### Breaking changes

* mirage-crypto: Poly1305 API now uses string (mirage/mirage-crypto#203 @hannesm)
* mirage-crypto: Poly1305 no longer has type alias "type mac = string"
  (mirage/mirage-crypto#232 @hannesm)
* mirage-crypto: the API uses string instead of cstruct (mirage/mirage-crypto#214 @reynir @hannesm)
* mirage-crypto: Hash module has been removed. Use digestif if you need hash
  functions (mirage/mirage-crypto#213 @hannesm)
* mirage-crypto: the Cipher_block and Cipher_stream modules have been removed,
  its contents is inlined:
  Mirage_crypto.Cipher_block.S -> Mirage_crypto.Block
  Mirage_crypto.Cipher_stream.S -> Mirage_crypto.Stream
  Mirage_crypto.Cipher_block.AES.CTR -> Mirage_crypto.AES.CTR
  (mirage/mirage-crypto#225 @hannesm, suggested in mirage/mirage-crypto#224 by @reynir)
* mirage-crypto-pk: s-expression conversions for private and public keys (Dh,
  Dsa, Rsa) have been removed. You can use PKCS8 for encoding and decoding
  `X509.{Private,Public}_key.{en,de}code_{der,pem}` (mirage/mirage-crypto#208 @hannesm)
* mirage-crypto-pk: in the API, Cstruct.t is no longer present. Instead,
  string is used (mirage/mirage-crypto#211 @reynir @hannesm)
* mirage-crypto-rng: the API uses string instead of Cstruct.t. A new function
  `generate_into : ?g -> bytes -> ?off:int -> int -> unit` is provided
  (mirage/mirage-crypto#212 @hannesm @reynir)
* mirage-crypto-ec: remove NIST P224 support (mirage/mirage-crypto#209 @hannesm @Firobe)
* mirage-crypto: in Uncommon.xor_into the arguments ~src_off and ~dst_off are
  required now (mirage/mirage-crypto#232 @hannesm), renamed to unsafe_xor_into
  (98f01b14f5ebf98ba0e7e9c2ba97ec518f90fddc)
* mirage-crypto-pk, mirage-crypto-rng: remove type alias "type bits = int"
  (mirage/mirage-crypto#236 @hannesm)

### Bugfixes

* mirage-crypto (32 bit systems): CCM with long adata (mirage/mirage-crypto#207 @reynir)
* mirage-crypto-ec: fix K_gen for bitlen mod 8 != 0 (reported in mirage/mirage-crypto#105 that
  P521 test vectors don't pass, re-reported mirage/mirage-crypto#228, fixed mirage/mirage-crypto#230 @Firobe)
* mirage-crypto-ec: zero out bytes allocated for Field_element.zero (reported
  mirleft/ocaml-x509#167, fixed mirage/mirage-crypto#226 @dinosaure)

### Data race free

* mirage-crypto (3DES): avoid global state in key derivation (mirage/mirage-crypto#223 @hannesm)
* mirage-crypto-rng: use atomic instead of reference to be domain-safe (mirage/mirage-crypto#221
  @dinosaure @reynir @hannesm)
* mirage-crypto, mirage-crypto-rng, mirage-crypto-pk, mirage-crypto-ec:
  avoid global buffers, use freshly allocated strings/bytes instead, avoids
  data races (mirage/mirage-crypto#186 mirage/mirage-crypto#219 @dinosaure @reynir @hannesm)

### Other changes

* mirage-crypto: add {de,en}crypt_into functions (and unsafe variants) to allow
  less buffer allocations (mirage/mirage-crypto#231 @hannesm)
* mirage-crypto-rng-miou: new package which adds rng support with miou
  (mirage/mirage-crypto#227 @dinosaure)
* PERFORMANCE mirage-crypto: ChaCha20/Poly1305 use string instead of Cstruct.t,
  ChaCha20 interface unchanged, performance improvement roughly 2x
  (mirage/mirage-crypto#203 @hannesm @reynir)
* mirage-crypto-ec, mirage-crypto-pk, mirage-crypto-rng: use digestif for
  hashes (mirage/mirage-crypto#212 mirage/mirage-crypto#215 @reynir @hannesm)
* mirage-crypto-rng: use a set for entropy sources instead of a list
  (mirage/mirage-crypto#218 @hannesm)
* mirage-crypto-rng-mirage: provide a module type S (for use instead of
  mirage-random in mirage) (mirage/mirage-crypto#234 @hannesm)

### CODE REVIEW

**Comment by hannesm** [comment]
This is a major release, breaking API changes (see https://github.com/ocaml/opam-repository/pull/26045 for already restricted upper bounds). Subsequently we'll release reverse dependencies that are compatible with mirage-crypto 1.0.0. It is a big milestone that we'll announce soon [tm]. :)

After all, there's no longer a dependency onto `cstruct`.

**Comment by hannesm** [comment]
Please note that arm64 test failures are fine for now (see https://github.com/mirage/mirage-crypto/issues/216 for further details, maybe someone even has an idea for what to do on arm64 platforms).

File: packages/mirage-crypto-rng-miou-unix/mirage-crypto-rng-miou-unix.1.0.0/opam [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+         ["dune" "runtest" "-p" name "-j" jobs] {with-test} ]
+
+depends: [
+  "ocaml" {>= "5.0.0"}
+  "dune" {>= "2.7"}
+  "miou" {>= "0.2.0"}
+  "logs"
+  "mirage-crypto-rng" {=version}
+  "duration"
+  "mtime"
```
Comment (mseri) [inline packages/mirage-crypto-rng-miou-unix/mirage-crypto-rng-miou-unix.1.0.0/opam:22]:
```suggestion
  "mtime" {>= "1.0.0"}
```

due to
```

#=== ERROR while compiling mirage-crypto-rng-miou-unix.1.0.0 ==================#
# context              2.2.0~rc1~dev | linux/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/mirage/mirage-crypto/releases/download/v1.0.0/mirage-crypto-1.0.0.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/mirage-crypto-rng-miou-unix.1.0.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p mirage-crypto-rng-miou-unix -j 255
# exit-code            1
# env-file             ~/.opam/log/mirage-crypto-rng-miou-unix-7-56dac7.env
# output-file          ~/.opam/log/mirage-crypto-rng-miou-unix-7-56dac7.out
### output ###
# File "rng/miou/dune", line 4, characters 100-114:
# 4 |  (libraries miou miou.unix mirage-crypto mirage-crypto-rng mirage-crypto-rng.unix digestif duration mtime.clock.os logs)
#                                                                                                         ^^^^^^^^^^^^^^
# Error: Library "mtime.clock.os" not found.
# -> required by library "mirage-crypto-rng-miou-unix" in
#    _build/default/rng/miou
# -> required by _build/default/META.mirage-crypto-rng-miou-unix
# -> required by _build/install/default/lib/mirage-crypto-rng-miou-unix/META
# -> required by _build/default/mirage-crypto-rng-miou-unix.install
# -> required by alias install

```

**Comment by mseri** [comment]
Thanks

**Pull request merged**